source/common/http/solo_filter_utility.h

```diff
@@ -38,5 +38,35 @@ class SoloFilterUtility {
   resolveClusterName(StreamFilterCallbacks *filter_callbacks);
 };
 
+class PerFilterConfigUtilBase {
+protected:
+  PerFilterConfigUtilBase(const std::string &filter_name)
+      : filter_name_(filter_name) {}
+  const Router::RouteSpecificFilterConfig *
+  getPerFilterBaseConfig(StreamFilterCallbacks &filter_callbacks);
+
+private:
+  const std::string &filter_name_;
+  Router::RouteConstSharedPtr route_info_{};
+};
+
+template <class ConfigType>
+class PerFilterConfigUtil : public PerFilterConfigUtilBase {
+
+  static_assert(
+      std::is_base_of<Router::RouteSpecificFilterConfig, ConfigType>::value,
+      "ConfigType must be a subclass of Router::RouteSpecificFilterConfig");
+
+public:
+  PerFilterConfigUtil(const std::string &filter_name)
+      : PerFilterConfigUtilBase(filter_name) {}
+
+  const ConfigType *
+  getPerFilterConfig(StreamFilterCallbacks &filter_callbacks) {
+    return dynamic_cast<const ConfigType *>(
+        getPerFilterBaseConfig(filter_callbacks));
+  }
+};
```

Comment:
I am not sure whether `PerFilterConfigUtilBase` is required.
It does not have any public member function, so users are likely to have to only access the derived class directly.
Consider just making `getPerFilterBaseConfig()` a private member function of `PerFilterConfigUtil<>`.

Reply:
do you think it will work  in the .cc file?
i did it this way to have the most of the code in .cc file to avoid too much code in header